### PR TITLE
v0.38.x: backport of new content on spec/p2p

### PR DESCRIPTION
**Pull request by cason**
This PR backports 3 PRs merged into the `main` branch: #714, #851, and #966.

Note to reviewers: there is no new content added here, this is just a manual backport of multiple PRs already merged into main.

Addresses #598

### CODE REVIEW

File: p2p/README.md [hunk 1]
```diff
@@ -4,7 +4,7 @@ The p2p package provides an abstraction around peer-to-peer communication.
...
 Docs:
 
-- [Connection](https://github.com/cometbft/cometbft/blob/v0.38.x/spec/p2p/connection.md) for details on how connections and multiplexing work
-- [Peer](https://github.com/cometbft/cometbft/blob/v0.38.x/spec/p2p/node.md) for details on peer ID, handshakes, and peer exchange
-- [Node](https://github.com/cometbft/cometbft/blob/v0.38.x/spec/p2p/node.md) for details about different types of nodes and how they should work
-- [Config](https://github.com/cometbft/cometbft/blob/v0.38.x/spec/p2p/config.md) for details on some config option
+- [Connection](https://github.com/cometbft/cometbft/blob/v0.38.x/spec/p2p/legacy-docs/connection.md) for details on how connections and multiplexing work
+- [Peer](https://github.com/cometbft/cometbft/blob/v0.38.x/spec/p2p/legacy-docs/node.md) for details on peer ID, handshakes, and peer exchange
+- [Node](https://github.com/cometbft/cometbft/blob/v0.38.x/spec/p2p/legacy-docs/node.md) for details about different types of nodes and how they should work
+- [Config](https://github.com/cometbft/cometbft/blob/v0.38.x/spec/p2p/legacy-docs/config.md) for details on some config option
```
Comment (thanethomson) [inline p2p/README.md:10]: Same as https://github.com/cometbft/cometbft/pull/1004#discussion_r1235564777

Comment (cason) [reply p2p/README.md]: Removed full links in https://github.com/cometbft/cometbft/pull/1005/commits/58777000c2aeb8fa09f3a40b52da740097a7aeac

File: spec/p2p/implementation/README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+The documentation was [produced](https://github.com/tendermint/tendermint/pull/9348)
+using the `v0.34.*` releases
+and the branch [`v0.34.x`](https://github.com/cometbft/cometbft/tree/v0.34.x)
+of this repository as reference.
+As there were no substancial changes in the p2p implementation, the
+documentation also applies to the releases `v0.37.*` and `v0.38.*` [^v35].
+
+[^v35]: The releases `v0.35.*` and `v0.36.*`, which included a major
+  refactoring of the p2p layer implementation, were [discontinued][v35postmorten].
```
Comment (lasarojc) [inline spec/p2p/implementation/README.md:14]: "discontinued" gives the impression that they didn't receive updates. I suggest a stronger wording here.

```suggestion
  refactoring of the p2p layer implementation, were [retired][v35postmorten].
```

File: spec/p2p/implementation/switch.md [hunk 3]
```diff
@@ -50,11 +50,11 @@ The `DialPeersAsync` method receives a list of peer addresses (strings)
 and dials all of them in parallel.
 It is invoked in two situations:
 
-- In the [setup](https://github.com/cometbft/cometbft/blob/29c5a062d23aaef653f11195db55c45cd9e02715/node/node.go#L985) of a node, to establish connections with every configured
-  persistent peer
+- In the [setup](https://github.com/cometbft/cometbft/blob/v0.34.x/node/node.go#L987)
```
Comment (lasarojc) [inline spec/p2p/implementation/switch.md:53]: v0.34.x could still be updated and these links be broken. What is the guideline here? Should we incur in the risk of broken links or point to specific patch versions?

**Review by lasarojc — APPROVED** [review]
Good stuff. A couple of minor comments but non-blocking. LGTM.

File: spec/p2p/implementation/switch.md (hunk 3 above)
Comment (cason) [reply spec/p2p/implementation/switch.md:53]: So, the `implementation/README.md` specifies that this documentation is based on `v0.34.x` branch. I've tried to check whether replacing links to the `main` branch could to the work, but it doesn't. So I would rather maintain references to `v0.34.x` here, until we find a way to describe the implementation on all branches consistently.

**Comment by cason** [comment]
> Good stuff. A couple of minor comments, which I should have made on the main branch. LGTM.

I also realized, from your reviews and from mine, that there are some small changes to make, but I would rather open another PR against `main` with them, then to backport this new PR to other branches.

**Review by sergio-mena — APPROVED** [review]
Didn't read through the text. I started but realized I've already reviewed this (it's a backport -- let me know if there's some text not present in `main` that I should read)

**Comment by cason** [comment]
@sergio-mena, sorry, I have updated the description. The changes are really minimal:

```
$ git checkout cason/p2p-backport-to-0.38
$ git diff --stat main -- spec/p2p/ spec/abci/ p2p/README.md
 p2p/README.md                              |  8 ++++----
 spec/abci/README.md                        |  2 +-
 spec/abci/abci++_app_requirements.md       | 35 +++++++++++------------------------
 spec/abci/abci++_client_server.md          |  2 +-
 spec/abci/abci++_methods.md                | 12 ++++--------
 spec/p2p/legacy-docs/messages/consensus.md |  2 +-
 spec/p2p/legacy-docs/messages/mempool.md   |  2 +-
 7 files changed, 23 insertions(+), 40 deletions(-)
```

**Comment by cason** [comment]
The other backport of this same content, but to `v0.37.x`: https://github.com/cometbft/cometbft/pull/1004

**Comment by cason** [comment]
Any blocker for merging this PR?

**Comment by sergio-mena** [comment]
> I have updated the description. The changes are really minimal

I've reviewed those changes locally and LGTM, except for a minor comment that I just added.

For me it looks good to go

**Review by sergio-mena — APPROVED** [review]

File: spec/p2p/legacy-docs/messages/consensus.md [hunk 4]
```diff
@@ -24,21 +24,21 @@ next block in the blockchain should be.
...
 |----------|----------------------------------------------------|----------------------------------------|--------------|
-| proposal | [Proposal](../../core/data_structures.md#proposal) | Proposed Block to come to consensus on | 1            |
+| proposal | [Proposal](../../../core/data_structures.md#proposal) | Proposed Block to come to consensus on | 1            |
 
 ### Vote
 
 Vote is sent to vote for some block (or to inform others that a process does not vote in the
 current round). Vote is defined in the
-[Blockchain](https://github.com/cometbft/cometbft/blob/v0.38.x/spec/core/data_structures.md#blockidd)
+[Blockchain](../../..//core/data_structures.md#blockidd)
```
[the remark this answers is not in the document]
Comment (cason) [reply spec/p2p/legacy-docs/messages/consensus.md:33]:
```suggestion
[Blockchain](../../../core/data_structures.md#blockidd)
```

[the remark this answers is not in the document]
Comment (cason) [reply spec/p2p/legacy-docs/messages/consensus.md]: Typo, fixed, thanks.

**Pull request merged**